### PR TITLE
Fairuz editable tasks

### LINKED ISSUE #45: As a lazy user, I can edit or delete tasks I've added before
So that I save time and energy.

### PR DESCRIPTION
**Pull request by skyaxe97**
Fixes #45

### CODE REVIEW

File: src/main/java/seedu/lifeasier/commands/DeleteTaskCommand.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                throw new ParserException();
+            }
+            logger.log(Level.INFO, "Start of DeleteTaskCommand");
+            ui.showSelectTaskToDelete(type);
+            printMatchingTasks(tasks, ui, type, name);
+            int userTaskChoice = ui.readSingleIntInput() - 1;
+            checkForIndexOutOfBounds(tasks, userTaskChoice);
+            deleteTask(tasks, ui, userTaskChoice);
+            storage.saveTasks();
+            ui.showDeleteConfirmationMessage();
```
Comment (dojh111) [inline src/main/java/seedu/lifeasier/commands/DeleteTaskCommand.java]: Maybe can just add some white spaces to separate the different logic to improve the code readability/aesthetics abit! Same for the other edit commands below.

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 2]
```diff
@@ -153,6 +168,117 @@ private Command parseAddNotesCommand(String input) {
...
+                    throw new ParserException();
+                }
+
+                String date = input.substring(lastIndexOfDateCommand, firstIndexOfTimeCommand).trim();
+                String startTime = input.substring(lastIndexOfTimeCommand, firstIndexOfToCommand).trim();
+                String endTime =  input.substring(lastIndexOfToCommand).trim();
+                LocalDateTime start = LocalDateTime.parse(date + " " + startTime, DATE_TIME_FORMATTER);
+                LocalDateTime end = LocalDateTime.parse(date + " " + endTime, DATE_TIME_FORMATTER);
+                times[0] = start;
+                times[1] = end;
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: Could consider using constants for these:
```suggestion
                times[INDEX_START] = start;
                times[INDEX_END] = end;
```

where INDEX_START = 0 and INDEX_END = 1

Comment (daniellimzj) [reply src/main/java/seedu/lifeasier/parser/Parser.java]: same for getting the "by" of the deadline as well!

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 3]
```diff
@@ -153,6 +168,117 @@ private Command parseAddNotesCommand(String input) {
...
+
+            int firstIndexOfNameCommand = input.indexOf(PARAM_NAME);
+            if (firstIndexOfNameCommand == -1) {
+                type = input.substring(lastIndexOfTypeCommand).trim();
+                name = "";
+            } else {
+                int lastIndexOfNameCommand = input.indexOf(PARAM_NAME) + PARAM_NAME.length();
+                type = input.substring(lastIndexOfTypeCommand, firstIndexOfNameCommand).trim();
+                if (!type.equals(PARAM_DEADLINE) && !type.equals(PARAM_EVENT)
+                        && !type.equals(PARAM_LESSON)) {
```
Comment (dojh111) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: Contents of if conditional may be extracted into another method to return true or false

File: src/main/java/seedu/lifeasier/tasks/TaskList.java [hunk 4]
```diff
@@ -76,7 +83,84 @@ public static void addDeadline(String description, LocalDateTime by) {
...
+    }
+
+    public void editLessonTime(int index, Ui ui) throws ParserException {
+        Parser parser = new Parser();
+        LocalDateTime[] times = parser.parseNewTimeInput(ui, ui.readCommand(), 2);
+        if (times[0] == null) {
+            throw new ParserException();
+        }
+        getTask(index).setStart(times[0]);
+        getTask(index).setEnd(times[1]);
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/tasks/TaskList.java]: Same comment as the earlier one!

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 5]
```diff
@@ -153,6 +168,117 @@ private Command parseAddNotesCommand(String input) {
...
+
+            case (2):
+                int firstIndexOfDateCommand = input.indexOf(PARAM_DATE);
+                int lastIndexOfDateCommand = firstIndexOfDateCommand + PARAM_DATE.length();
+                int firstIndexOfTimeCommand = input.indexOf(PARAM_TIME);
+                int lastIndexOfTimeCommand = firstIndexOfTimeCommand + PARAM_TIME.length();
+                int firstIndexOfToCommand = input.indexOf(PARAM_TO);
+                int lastIndexOfToCommand = firstIndexOfToCommand + PARAM_TO.length();
+
+                if (firstIndexOfDateCommand == -1 || firstIndexOfTimeCommand == -1 || firstIndexOfToCommand == -1) {
```
Comment (dojh111) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: Same for this if conditional statement. Can be extracted into a method if you can find a name for it

File: src/main/java/seedu/lifeasier/ui/Ui.java [hunk 6]
```diff
@@ -92,10 +105,98 @@ public String readCommand() {
...
+        case (PARAM_EVENT):
+            System.out.println("Please input your new Event name");
+            break;
+
+        default:
+            break;
+        }
+    }
+
+    public void showEditableParameters(String type) {
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/ui/Ui.java]: Could you consider renaming this to include "Message" for consistency?
```suggestion
    public void showEditableParametersMessage(String type) {
```

**Review by daniellimzj — COMMENTED** [review]
Just a few minor issues!
It would also be good to try and throw in some assertions

File: src/main/java/seedu/lifeasier/tasks/TaskList.java [hunk 7]
```diff
@@ -76,7 +83,84 @@ public static void addDeadline(String description, LocalDateTime by) {
...
+    }
+
+    public void printMatchingTasks(Ui ui, String type, String description) throws TaskNotFoundException {
+
+        logger.log(Level.INFO, "Start of printing all matching " + type);
+        indexOfLastMatch = 0;
+        boolean noMatches = true;
+        for (int i = 0; i < getTaskCount(); i++) {
+            if (getTask(i).getType().equals(type)
+                    && getTask(i).getDescription().contains(description)) {
```
Comment (dojh111) [inline src/main/java/seedu/lifeasier/tasks/TaskList.java]: Same here, for the if conditional

File: src/main/java/seedu/lifeasier/tasks/TaskList.java [hunk 8]
```diff
@@ -76,7 +83,84 @@ public static void addDeadline(String description, LocalDateTime by) {
...
+            taskList.remove(index);
+            taskCount--;
+        } catch (IndexOutOfBoundsException e) {
+            ui.showInvalidNumberMessage();
+        } catch (NumberFormatException e) {
+            ui.showNumberFormatMessage();
+        }
+    }
+
+    public void printMatchingTasks(Ui ui, String type, String description) throws TaskNotFoundException {
```
Comment (dojh111) [inline src/main/java/seedu/lifeasier/tasks/TaskList.java]: Maybe this method can go into Ui, will need to check with the rest

**Review by dojh111 — COMMENTED** [review]
Looks pretty good to me, very clean. Only need to extract out the few longer if conditionals to make it more readable. Good work!

**Pull request merged**